libcaf_net/caf/net/datagram_transport.hpp

```diff
@@ -0,0 +1,327 @@
+/******************************************************************************
+ *                       ____    _    _____                                   *
+ *                      / ___|  / \  |  ___|    C++                           *
+ *                     | |     / _ \ | |_       Actor                         *
+ *                     | |___ / ___ \|  _|      Framework                     *
+ *                      \____/_/   \_|_|                                      *
+ *                                                                            *
+ * Copyright 2011-2019 Dominik Charousset                                     *
+ *                                                                            *
+ * Distributed under the terms and conditions of the BSD 3-Clause License or  *
+ * (at your option) under the terms and conditions of the Boost Software      *
+ * License 1.0. See accompanying files LICENSE and LICENSE_ALTERNATIVE.       *
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#pragma once
+
+#include <deque>
+#include <unordered_map>
+
+#include "caf/byte.hpp"
+#include "caf/detail/socket_sys_aliases.hpp"
+#include "caf/detail/socket_sys_includes.hpp"
+#include "caf/error.hpp"
+#include "caf/fwd.hpp"
+#include "caf/ip_endpoint.hpp"
+#include "caf/logger.hpp"
+#include "caf/net/defaults.hpp"
+#include "caf/net/endpoint_manager.hpp"
+#include "caf/net/fwd.hpp"
+#include "caf/net/receive_policy.hpp"
+#include "caf/net/transport_worker_dispatcher.hpp"
+#include "caf/net/udp_datagram_socket.hpp"
+#include "caf/sec.hpp"
+#include "caf/span.hpp"
+#include "caf/variant.hpp"
+
+namespace caf {
+namespace net {
+
+/// Implements a udp_transport policy that manages a datagram socket.
+template <class Factory>
+class datagram_transport {
+public:
+  // -- member types -----------------------------------------------------------
+
+  using id_type = ip_endpoint;
+
+  using buffer_type = std::vector<byte>;
+
+  using buffer_cache_type = std::vector<buffer_type>;
+
+  using factory_type = Factory;
+
+  using transport_type = datagram_transport<Factory>;
+
+  using dispatcher_type = transport_worker_dispatcher<transport_type, id_type>;
+
+  using application_type = typename Factory::application_type;
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  datagram_transport(udp_datagram_socket handle, factory_type factory)
+    : dispatcher_(*this, std::move(factory)),
+      handle_(handle),
+      max_consecutive_reads_(0),
+      read_threshold_(1024),
+      max_(1024),
```

Comment:
Is this a hard maximum? In any case, I think the magic numbers should at least go into some header with a proper name.